### PR TITLE
refactor: cleanup `PigeonholeSort`

### PR DESCRIPTION
**Pull request by alxkm**
This PR adds missed test.

Improve code structure, this should add more readability.

Replaced `ArrayList<Integer>[] pigeonHole = new ArrayList[numOfPigeonholes];` with `List<List<Integer>> pigeonHoles = new ArrayList<>(maxElement + 1);`. 
Because such constructions in Java is generally not considered good practice.

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/sorts/PigeonholeSort.java [hunk 1]
```diff
@@ -1,55 +1,88 @@
...
+    /**
+     * Sorts the given array using the pigeonhole sort algorithm.
+     *
+     * @param array the array to be sorted
+     * @throws IllegalArgumentException if any negative integers are found
+     * @return the sorted array
+     */
+    public static int[] sort(int[] array) {
+        if (array.length < 2) {
+            return array;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/PigeonholeSort.java:20]:
```suggestion
        checkForNegativeInput(array);

        if (array.length == 0) {
            return array;
```

Two things:
- the check for negative numbers has to happen first,
- it seems enough to check if the array is empty.

File: src/main/java/com/thealgorithms/sorts/PigeonholeSort.java [hunk 2]
```diff
@@ -1,55 +1,88 @@
...
         }
 
-        int numOfPigeonholes = 1 + maxElement;
-        ArrayList<Integer>[] pigeonHole = new ArrayList[numOfPigeonholes];
+        checkForNegativeInput(array);
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/PigeonholeSort.java:23]:
```suggestion
```

File: src/main/java/com/thealgorithms/sorts/PigeonholeSort.java [hunk 3]
```diff
@@ -1,55 +1,88 @@
...
-                array[k] = elements;
-                k = k + 1;
+    /**
+     * Checks if the array contains any negative integers.
+     *
+     * @param array the array to be checked
+     * @throws IllegalArgumentException if any negative integers are found
+     */
+    private static void checkForNegativeInput(int[] array) {
+        for (int number : array) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/PigeonholeSort.java:41]:
```suggestion
        for (final int number : array) {
```

File: src/main/java/com/thealgorithms/sorts/PigeonholeSort.java [hunk 4]
```diff
@@ -1,55 +1,88 @@
...
+            if (number < 0) {
+                throw new IllegalArgumentException("Array contains non-positive integers.");
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/PigeonholeSort.java:43]:
```suggestion
                throw new IllegalArgumentException("Array contains negative integers.");
```

File: src/main/java/com/thealgorithms/sorts/PigeonholeSort.java [hunk 5]
```diff
@@ -1,55 +1,88 @@
...
+    /**
+     * Collects sorted elements from the pigeonholes back into the array.
+     *
+     * @param array the array to be sorted
+     * @param pigeonHoles the populated pigeonholes
+     */
+    private static void collectFromPigeonHoles(int[] array, List<List<Integer>> pigeonHoles) {
+        int index = 0;
+        for (List<Integer> pigeonHole : pigeonHoles) {
+            for (int element : pigeonHole) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/PigeonholeSort.java:83]:
```suggestion
        for (final var pigeonHole : pigeonHoles) {
            for (final int element : pigeonHole) {
```

File: src/test/java/com/thealgorithms/sorts/PigeonholeSortTest.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    private static Stream<Arguments> provideArraysForPigeonholeSort() {
+        return Stream.of(Arguments.of(new int[] {}, new int[] {}), Arguments.of(new int[] {4}, new int[] {4}), Arguments.of(new int[] {6, 1, 99, 27, 15, 23, 36}, new int[] {1, 6, 15, 23, 27, 36, 99}), Arguments.of(new int[] {6, 1, 27, 15, 23, 27, 36, 23}, new int[] {1, 6, 15, 23, 23, 27, 27, 36}),
+            Arguments.of(new int[] {5, 5, 5, 5, 5}, new int[] {5, 5, 5, 5, 5}), Arguments.of(new int[] {1, 2, 3, 4, 5}, new int[] {1, 2, 3, 4, 5}), Arguments.of(new int[] {5, 4, 3, 2, 1}, new int[] {1, 2, 3, 4, 5}));
+    }
+
+    @Test
+    public void testWithNegativeNumbers() {
+        assertThrows(IllegalArgumentException.class, () -> PigeonholeSort.sort(new int[] {3, 1, 4, 1, 5, -9}));
```
Comment (vil02) [inline src/test/java/com/thealgorithms/sorts/PigeonholeSortTest.java:28]:
```suggestion
        assertThrows(IllegalArgumentException.class, () -> PigeonholeSort.sort(new int[] {3, 1, 4, 1, 5, -9}));
        assertThrows(IllegalArgumentException.class, () -> PigeonholeSort.sort(new int[] {-1}));
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/sorts/PigeonholeSort.java (hunk 1 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/sorts/PigeonholeSort.java]: Thanks.  Thanks. Input checking for empty array will be more clear than checking for exact size.

Is this way to check negative input first is better from code clarity and readability perspective? To highlight first that it doesn't work with negative numbers?

**Review by vil02 — APPROVED** [review]

**Pull request merged**